File: ruby/addressable.rb
class Addressable
	attr_accessor :name

	@@MAX_ADDR = 2 ** 64

	def initialize()
		@mem = {}
		@name = "???"
	end
	
	def read(pos, size)
		res = []
		for i in 0..size-1 do
			res.push(@mem.fetch((pos + i) % @@MAX_ADDR, 0))
		end
		res
	end
	
	def read_bit_array(pos, size)
		byte_arr = read(pos, size)
		return byte_arr.flat_map{|byte|
			[byte[0], byte[1], byte[2], byte[3],
			 byte[4], byte[5], byte[6], byte[7]]}
	end
	
	def write(pos, data)
		for i in 0..data.length-1 do
			@mem[(pos + i) % @@MAX_ADDR] = data[i]
		end
	end
	
	def write_bit_array(pos, bitarray)
		raise "Bit array does not map to byte array" unless bitarray.length % 8 == 0
		byte_arr = bitarray.each_slice(8)
		for i in 0..(byte_arr.length-1)
			item = byte_arr[i]
			byte = item[0] + 2 * item[1] + 4 * item[2] + 8 * item[3] +
				16 * item[4] + 32 * item[5] + 64 * item[6] + 127 * item[7]
			write(pos + i, [byte])
		end
	end
	
	def read_int(pos, size)
		read(pos, size).pack("C*").unpack(pack_scheme(size))[0]
	end
	
	def read_signed(pos, size)
		read_int(pos, size) - 2 ** (8 * size)
	end

	def pack_scheme(size)
		case size
		when 1
			return "C"
		when 2
			return "S<"
		when 4
			return "L<"
		when 8
			return "Q<"
		else
			raise "bad size: %d" % size
		end
	end
end

File: python/addressable.py
import struct


class Addressable:
    MAX_ADDR = 2 ** 64

    def __init__(self):
        self.mem = {}
        self.name = "???"

    def read(self, pos, size):
        res = []
        for i in range(size):
            res.push(self.mem.get((pos + i) % self.MAX_ADDR, 0))

        return res

    def read_bit_array(self, pos, size):
        byte_arr = self.read(pos, size)
        res = []
        for byte in byte_arr:
            res.append(byte & 0b00000001)
            res.append((byte & 0b00000010) >> 1)
            res.append((byte & 0b00000100) >> 2)
            res.append((byte & 0b00001000) >> 3)
            res.append((byte & 0b00010000) >> 4)
            res.append((byte & 0b00100000) >> 5)
            res.append((byte & 0b01000000) >> 6)
            res.append((byte & 0b10000000) >> 7)

        return res

    def write(self, pos, data):
        for i in len(data):
            self.mem[(pos + i) % self.MAX_ADDR] = data[i]

    def write_bit_array(self, pos, bitarray):
        if len(bitarray) % 8 != 0:
            raise "Bit array does not map to byte array"

        byte_arr = []

        for byte_start in bitarray[::8]:
            byte = bitarray[byte_start]
            byte += bitarray[byte_start + 1] << 1
            byte += bitarray[byte_start + 2] << 2
            byte += bitarray[byte_start + 3] << 3
            byte += bitarray[byte_start + 4] << 4
            byte += bitarray[byte_start + 5] << 5
            byte += bitarray[byte_start + 6] << 6
            byte += bitarray[byte_start + 7] << 7

            byte_arr.append(byte)

        self.write(pos, byte_arr)

    def read_int(self, pos, size):
        byte_arr = self.read(pos, size)
        string = struct.pack("B" * size, byte_arr)
        return struct.unpack(string, self.pack_scheme(size))[0]

    def read_signed(self, pos, size):
        byte_arr = self.read(pos, size)
        string = struct.pack("B" * size, byte_arr)
        return struct.unpack(string, self.pack_scheme(size).lower())[0]

    def pack_scheme(size):
        if size == 1:
            return "B"
        elif size == 2:
            return "<H"
        elif size == 4:
            return "<I"
        elif size == 8:
            return "<Q"
        else:
            raise "bad size: %d" % size
